src/effects/audiounits/AudioUnitEffect.cpp

```diff
@@ -203,57 +202,107 @@ bool AudioUnitEffect::SupportsAutomation() const
    return supports;
 }
 
-class AudioUnitInstance : public StatefulPerTrackEffect::Instance
+class AudioUnitInstance : public PerTrackEffect::Instance
    , public AudioUnitWrapper
+   , EffectInstanceWithSampleRate
 {
 public:
-   AudioUnitInstance(StatefulPerTrackEffect &effect,
-      AudioComponent component, Parameters &parameters);
-   ~AudioUnitInstance() override;
+   AudioUnitInstance(const PerTrackEffect &effect,
+      AudioComponent component, Parameters &parameters,
+      const wxString &identifier,
+      unsigned audioIns, unsigned audioOuts, bool useLatency);
 
    void EventListener(const AudioUnitEvent *inEvent,
       AudioUnitParameterValue inParameterValue);
+
+private:
+   size_t InitialBlockSize() const;
+   sampleCount GetLatency(const EffectSettings &settings) override;
+
+   size_t GetBlockSize() const override;
+   size_t SetBlockSize(size_t maxBlockSize) override;
+
+   bool ProcessInitialize(EffectSettings &settings,
+      sampleCount totalLen, ChannelNames chanMap) override;
+   bool ProcessFinalize() override;
+   size_t ProcessBlock(EffectSettings &settings,
+      const float *const *inBlock, float *const *outBlock, size_t blockLen)
+      override;
+
+   bool RealtimeInitialize(EffectSettings &settings) override;
+   bool RealtimeAddProcessor(EffectSettings &settings,
+      unsigned numChannels, float sampleRate) override;
+   bool RealtimeFinalize(EffectSettings &settings) noexcept override;
+   bool RealtimeSuspend() override;
+   bool RealtimeResume() override;
+   bool RealtimeProcessStart(EffectSettings &settings) override;
+   size_t RealtimeProcess(size_t group, EffectSettings &settings,
+      const float *const *inbuf, float *const *outbuf, size_t numSamples)
+      override;
+   bool RealtimeProcessEnd(EffectSettings &settings) noexcept override;
+
+   bool SetRateAndChannels();
+
+   static OSStatus RenderCallback(void *inRefCon,
+      AudioUnitRenderActionFlags *inActionFlags,
+      const AudioTimeStamp *inTimeStamp,
+      UInt32 inBusNumber, UInt32 inNumFrames, AudioBufferList *ioData);
+   OSStatus Render(AudioUnitRenderActionFlags *inActionFlags,
+      const AudioTimeStamp *inTimeStamp,
+      UInt32 inBusNumber, UInt32 inNumFrames, AudioBufferList *ioData);
+
+   bool BypassEffect(bool bypass);
+
+private:
+   //! Whether the master instance is now allocated to a group number
+   bool mRecruited{ false };
+   std::vector<std::unique_ptr<AudioUnitInstance>> mSlaves;
```

Comment:
I am a bit surprised to see that an Instance can have slaves. I thought the main point of having instances was to replace the slaves in the effect, externalizing them - so that what was once a slave, would now be an instance (and the first instance would be the "master")  maybe I am missing something here.

Reply:
I thought earlier we would need to externalize spaces as objects but decided against it.

Wahwah and Reverb have slaves implemented differently, not as occurrences of the instance class.

It doesn’t matter.  RealtimeAddProcessor and Process and Finalize are sufficient abstract interface on just one object.